File: src/IntersectionTwoLinked Lists.py
class ListNode(object):
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        if not headA or not headB:
            return
        l1 = headA
        l2 = headB
        c = 0
        while (l1 or l2) and c <= 2:
            if l1 == l2:
                return l1
            l1 = l1.next
            l2 = l2.next
            if not l1:
                l1 = headB
                c += 1
            if not l2:
                l2 = headA
                c += 1


sol = Solution()
l1 = ListNode(1)
l1.next = ListNode(2)
l1.next.next = ListNode(3)
l1.next.next.next = ListNode(4)

l2 = ListNode(0)
l2.next = l1.next.next

print(sol.getIntersectionNode(l2, l1).val)

File: src/IntersectionTwoLinkedLists.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }

    ListNode(int x, ListNode n) {
        this(x);
        this.next = n;
    }
}

public class IntersectionTwoLinkedLists {
    private int lenOf(ListNode head) {
        int len = 0;
        while (head != null) {
            len += 1;
            head = head.next;
        }
        return len;
    }

    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        //boundary check
        if (headA == null || headB == null) return null;

        ListNode a = headA;
        ListNode b = headB;

        //if a & b have different len, then we will stop the loop after second iteration
        while (a != b) {
            //for the end of first iteration, we just reset the pointer to the head of another linkedlist
            a = a == null ? headB : a.next;
            b = b == null ? headA : b.next;
        }
        return a;
    }
}
